python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -0,0 +1,186 @@
+
+    def add_output(self, ref: RefBundle) -> None:
+        self.outqueue.append(ref)
+        self.num_completed_tasks += 1
+        if self.progress_bar:
+            self.progress_bar.update(1)
+
+    def refresh_progress_bar(self) -> None:
+        if self.progress_bar:
+            queued = self.num_queued()
+            self.progress_bar.set_description(
+                f"{self.op.name}: {self.num_active_tasks()} active, {queued} queued"
+            )
+
+    def dispatch_next_task(self) -> None:
+        for i, inqueue in enumerate(self.inqueues):
+            if inqueue:
+                self.op.add_input(inqueue.pop(0), input_index=i)
+                return
+        assert False, "Nothing to dispatch"
+
+
+def build_streaming_topology(dag: PhysicalOperator) -> Topology:
+    """Build the streaming operator state topology for the given DAG.
+
+    This involves creating the operator state for each operator in the DAG,
+    registering it with this class, and wiring up the inqueues/outqueues of
+    dependent operator states.
+
+    Returns:
+        The topology dict holding the streaming execution state.
+    """
+
+    topology: Topology = {}
+
+    # DFS walk to wire up operator states.
+    def setup_state(op: PhysicalOperator) -> OpState:
+        if op in topology:
+            return topology[op]
+
+        # Create state if it doesn't exist.
+        op_state = OpState(op)
+        topology[op] = op_state
+
+        # Wire up the input outqueues to this op's inqueues.
+        for i, parent in enumerate(op.input_dependencies):
+            parent_state = setup_state(parent)
+            op_state.inqueues[i] = parent_state.outqueue
+
+        return op_state
+
+    setup_state(dag)
+
+    # Create the progress bars starting from the first operator to run.
+    i = 0
+    for op_state in list(topology.values())[::-1]:
+        if not isinstance(op_state.op, InputDataBuffer):
+            op_state.initialize_progress_bar(i)
+            i += 1
+
+    # TODO: fill out stats.
+    return topology, DatasetStats(stages={}, parent=None)
+
+
+def process_completed_tasks(topology: Topology) -> None:
+    """Process any newly completed tasks and update operator state."""
+    for op_state in topology.values():
+        op_state.refresh_progress_bar()
+
+    # Update active tasks.
+    active_tasks: Dict[ray.ObjectRef, PhysicalOperator] = {}
+
+    for op in topology:
+        for ref in op.get_work_refs():
+            active_tasks[ref] = op
+
+    # Process completed Ray tasks and notify operators.
+    if active_tasks:
+        completed, _ = ray.wait(
+            list(active_tasks),
+            num_returns=len(active_tasks),
+            fetch_local=False,
+            timeout=0.1,
+        )
+        for ref in completed:
+            op = active_tasks.pop(ref)
+            op.notify_work_completed(ref)
+
+    # Pull any operator outputs into the streaming op state.
+    for op, op_state in topology.items():
```

Comment:
Do we traverse the entire DAG instead of just joining `[active_tasks[ref] for ref in completed]` against `topology` and only processing those operator outputs to account for source operators like `InputDataBuffer` that might have outputs **without** newly completed work refs?

I'm asking along the lines of [this comment](https://github.com/ray-project/ray/pull/31443#discussion_r1062921732), wondering what the contract should be between the executors and operators on when the executor can expect the operator to (1) have new outputs, and (2) be done producing outputs. This current implementation seems to allow any operator to produce more outputs at any time, not only on observable operator events (e.g. work finishing for most operators, first pass for `InputDataBuffer`/source nodes, etc.)

Seems fine for now, but we can probably constrain this contract a bit more in the future, unless we think that the generic "always check every operator on every control loop step" treatment to be cleaner (and it will hopefully not have any perf implications for the control loop).

Reply:
Yes, that's right. I do think it's cleaner for now to always iterate over the entire topology. If it turns out to be a performance issue, we could optimize, but it's probably ok as long as pipelines are relatively simple (less than 10 operators).